:messages pane

tweekmonster opened the pull request:
This adds the `'messagepane'` (bool) option that causes `:messages` to display messages in a buffer.  Only messages that are kept in history are displayed in the pane.  The default message history has a limit of 200 entries, but the message pane keeps 1000 entries separately.

My motivation for doing this is keeping errors and spammy messages from interrupting me as I work.  I think this will be particularly useful for plugin development where the messages are intentionally verbose.

Todo:
- [x] ftplugin for the message pane that provides niceties, like jumping to an error source under the cursor.
- [x] Better implementation of "scroll to bottom" that can handle `'wrap'`.
- [x] Ignore errors from trying to modify the message pane buffer.
- [x] Vim script function for adding messages directly to the message pane.
- [x] Time stamped separator.  Could be used for simple profiling or separating groups of messages.
- [x] Tests
- [x] Rename to `msgbuf`, `message buffer`

Future Todo:
- Customizable history length.
- Specify message types to display in the pane (errors or echomsg).

tweekmonster committed:
Added message history "pane"

tweekmonster commented:
Demo: https://share.esdf.io/i3eC7pr0Dr/messages.html

justinmk commented:
See also https://github.com/neovim/neovim/pull/1802 which implements `:echomsg!` (notice the "bang") which is a way of adding message to the history and "maybe" printing it (but never triggering "Press Enter" prompt).

justinmk commented:
The approach that I assumed we would take would be to expose `:messages` as a list or something like that, so that functionality like this could be done in a plugin.

Or--perhaps better--expose each message as an API event. Then even collection of the messages is implemented outside of the C core. 

In general, we want to move away from implementing features in C if possible.

However, just IMO. Anything with good test coverage and reasonable no major objections by others, usually is worth including...

tweekmonster commented:
@justinmk Wish I saw #1802 sooner because it makes `messages.c` a little bit more understandable.

I tried to go the plugin route originally but the main thing I was missing was the colors.  I also felt that something like this would need to be as close as possible to the bare metal to keep it snappy.

tweekmonster committed:
Fill messages buffer if it didn't exist before.

tweekmonster committed:
Set winfixheight when opening a new window

tweekmonster committed:
Ignore e_modifiable if curbuf is the message pane buffer

tweekmonster committed:
Added msgpane() and msgpane_open() vimscript functions.

tweekmonster committed:
Added documentation

tweekmonster committed:
Reset message state even on empty strings.

tweekmonster committed:
scroll_to_bottom() works with wrapped lines.

tweekmonster committed:
Add messages from before the creation of the message pane buffer.

tweekmonster committed:
Changed filetype to `msgpane` and moved setting filetype.

tweekmonster committed:
Added msgpane scripts.

`gf` on an exception line lists functions that can be jumped to.

tweekmonster committed:
Changed p_msgpane to int.

- Some error about 'store to misaligned address'?
- Copied 'termguicolor' config in options.lua

tweekmonster committed:
Don't redraw if the buffer is updated but window isn't on screen.

tweekmonster committed:
Split the previous window when jumping to an error function.

tweekmonster committed:
Guard against duplicate history entries.

tweekmonster committed:
Ensure the last message is displayed in the command line.

tweekmonster committed:
Combine msgpane_attr with syntax_attr.

tweekmonster committed:
Added 'msgpane' to feature list.

tweekmonster committed:
Added os_timef()

tweekmonster committed:
Added message separator with timestamp.

tweekmonster commented:
Updated demo: https://share.esdf.io/ez62CPcIWI/messages.html

tweekmonster committed:
tests: Message pane

tweekmonster commented:
I'm still working on the tests, but I think this is ready for comments.

tweekmonster committed:
runtime: Recognize global functions.

tweekmonster committed:
runtime: Open function in existing buffer if available.

tweekmonster commented on test/functional/ui/message_pane_spec.lua:
```diff
@@ -0,0 +1,666 @@
+          /e\|a                                   |
+        ]])
+
+      end)
```
Reminder: Need to add tests for `msgpane()` and `msgpane_open()`

tweekmonster committed:
tests: Updated

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -5138,6 +5138,22 @@ msgpackparse({list})				   {Nvim} *msgpackparse()*
+		Open the message pane if it's not already displayed in a
+		window.  Only one message pane window is opened per tab.
+
+		Returns the message pane's buffer number or 0 is the pane
```
`is` -> `if`

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -7470,6 +7486,7 @@ localmap		Compiled with local mappings and abbr. |:map-local|
 mac			Macintosh version of Vim.
 macunix			Macintosh version of Vim, using Unix files (OS-X).
 menu			Compiled with support for |:menu|.
+msgpane			Compiled with support for 'messagepane'.
```
Do we need this line?

tweekmonster replied:
I copied a pattern.  I don't know if it's needed in the docs.  I wanted to have `has('msgpane')` so plugin developers could have some expectation of how messages are presented.

oni-link commented on src/nvim/eval.c:
```diff
@@ -13047,6 +13051,55 @@ static void f_msgpackparse(typval_T *argvars, typval_T *rettv)
+    attr = syn_id2attr(hl_id);
+  }
+
+  if (argvars[2].v_type != VAR_UNKNOWN) {
```
`argvars[2]` cannot always be checked. If only param `message` is used, we can only check up to `argvars[1]` and `argvars[1].v_type == VAR_UNKNOWN`.

oni-link commented on src/nvim/eval.c:
```diff
@@ -13047,6 +13051,55 @@ static void f_msgpackparse(typval_T *argvars, typval_T *rettv)
+///
+/// @param message Message to display.
+/// @param [highlight] The highlight group name.
+/// @param [open] Opens the message page if it's not already in a window.
```
`page` -> `pane`

mhinz commented:
General question: Shouldn't it be "window" instead of "pane"? This is not tmux after all. :-P

oni-link commented on src/nvim/eval.c:
```diff
@@ -13047,6 +13051,55 @@ static void f_msgpackparse(typval_T *argvars, typval_T *rettv)
+  if (argvars[1].v_type != VAR_UNKNOWN) {
+    char_u *attrName = get_tv_string(&argvars[1]);
+    int hl_id = syn_name2id(attrName);
+    attr = syn_id2attr(hl_id);
```
ASAN finds a problem with `:echo msgpane("asdf", 0,0)`. Parameter values not checked?
```c
==4831==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62d00023a3b8 at pc 0x0000016d5ac9 bp 0x7ffc8214cf40 sp 0x7ffc8214cf38
READ of size 4 at 0x62d00023a3b8 thread T0
    #0 0x16d5ac8 in syn_id2attr /home/oni-link/git/neovim/src/nvim/syntax.c:7235:15
    #1 0x89857c in f_msgpane /home/oni-link/git/neovim/src/nvim/eval.c:13071:12
    #2 0x794e63 in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7310:11
    #3 0x7a8da4 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7159:11
    #4 0x839119 in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4277:15
    #5 0x834c45 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3998:7
    #6 0x8318d9 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3850:7
    #7 0x82ce51 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3602:7
    #8 0x82c3e3 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3524:7
    #9 0x82b973 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3461:7
    #10 0x78d952 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3394:7
    #11 0x7eb20e in ex_echo /home/oni-link/git/neovim/src/nvim/eval.c:19415:9
    #12 0xa990d0 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2193:5
    #13 0xa77c17 in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:601:20
    #14 0x100c1fb in nv_colon /home/oni-link/git/neovim/src/nvim/normal.c:4489:18
    #15 0xff3743 in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1144:3
    #16 0x1684d10 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:56:26
    #17 0xfa7ab0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #18 0xd606c2 in main /home/oni-link/git/neovim/src/nvim/main.c:544:3
    #19 0x7f7cc90235af in __libc_start_main (/lib64/libc.so.6+0x205af)
    #20 0x445628 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x445628)

0x62d00023a3b8 is located 72 bytes to the left of 34056-byte region [0x62d00023a400,0x62d000242908)
allocated by thread T0 here:
    #0 0x4f8fb8 in realloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:77
    #1 0xe8385c in xrealloc /home/oni-link/git/neovim/src/nvim/memory.c:141:15
    #2 0xc84e33 in ga_grow /home/oni-link/git/neovim/src/nvim/garray.c:96:14
    #3 0x16c55ef in highlight_changed /home/oni-link/git/neovim/src/nvim/syntax.c:7376:3
    #4 0x1321b77 in update_screen /home/oni-link/git/neovim/src/nvim/screen.c:342:5
    #5 0x1069995 in normal_redraw /home/oni-link/git/neovim/src/nvim/normal.c:1263:5
    #6 0x106702c in normal_check /home/oni-link/git/neovim/src/nvim/normal.c:1346:5
    #7 0x1684723 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:20:35
    #8 0xfa7ab0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #9 0xd606c2 in main /home/oni-link/git/neovim/src/nvim/main.c:544:3
    #10 0x7f7cc90235af in __libc_start_main (/lib64/libc.so.6+0x205af)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/neovim/src/nvim/syntax.c:7235:15 in syn_id2attr
```

oni-link commented on src/nvim/message_pane.c:
```diff
@@ -0,0 +1,302 @@
+  }
+
+  msgpane_buf = buflist_new((char_u *)"nvim://messages", NULL, 1, BLN_LISTED);
+  msgpane_buf->b_messages = true;
```
`msgpane_buf` could be `NULL`.

tweekmonster replied:
Ah, because `buflist_new()` could return NULL, right?  (just woke up)

oni-link commented on src/nvim/message_pane.c:
```diff
@@ -0,0 +1,302 @@
+    if (win_split((int)p_cwh, WSP_BOT) == FAIL) {
+      beep_flush();
+      unblock_autocmds();
+      return false;
```
Missing `msg_silent--;`?

justinmk commented:
Perhaps the feature should be called"message buffer". It is a buffer, right? Does it appear in the buffer list (I would say yes)? Messages normally shows in "the pager".

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -5138,6 +5138,22 @@ msgpackparse({list})				   {Nvim} *msgpackparse()*
 			representing extension type. Second is 
 			|readfile()|-style list of strings.
 
+msgpane({message}[, {highlight}[, {open}]])		{Nvim} *msgpane()*
```
So is this a completely separate mechanism from `:messages`? I thought the msgpane would collect all `:echomsg` messages.

justinmk replied:
If msgpane collects echomsg, why is this function needed?

tweekmonster replied:
I was thinking from the perspective of debug logging for plugin development.  I thought it would be nice if a message could get into the buffer without overwriting what's displayed in the command line.  I also wanted a way to make sure an important message wouldn't go unnoticed.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -309,7 +309,7 @@ close_buffer (
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```
I already don't like all the special handling of buftype=terminal. Can't we avoid another special case?

tweekmonster replied:
I was actually thinking about having one variable like `buf->b_internal_type` which is an int for un-editable buffers.  `BUF_TERMINAL`, `BUF_MESSAGES`, etc.  If it was non-zero then the same rules would apply to all of them.

justinmk replied:
That sounds better, but avoiding special treatment would still be ideal.

tweekmonster replied:
The main thing I was going for by following the terminal checks is I didn't want the message buffer to be editable at all.  I think that if a user could change certain buffer settings, it could lead to some bad behavior from writing messages to a buffer that the user expects to be controlling.

justinmk replied:
I recommend not trying to solve that problem before it actually is proven to be a problem. Even for terminal buffers, I think we're doing it wrong.

If something weird happens because the user explicitly made something  weird happen: who cares? :)

tweekmonster replied:
I think I segfaulted when trying to edit the buffer and a message came in .  But that was in the early stage.  Doesn't look like the problem still exists.

I removed all the `b_messages` checks and remembered the actual reason: The line highlights need to match the buffer lines.  It was easier to do because `bufhl_add_hl()` requires a highlight ID which I couldn't get without touching `:echohl` and a bunch of other functions.  I also thought it'd be kind of messed up if the user thought they controlled a buffer and lines started disappearing because the messages are being truncated.

justinmk replied:
We can set nomodfiable on the buffer.

tweekmonster replied:
I'm going to look at removing the special treatment after I sort out how the lines are added to the buffer.  As it is now, if a user happens to add a line after the buffer reaches 1000 lines, the buffer will grow indefinitely.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -321,8 +321,9 @@ close_buffer (
-    // terminal buffers can only be wiped
+  if ((buf->terminal || buf->b_messages)
+      && (unload_buf || del_buf || wipe_buf)) {
+    // terminal and message pane buffers can only be wiped
```
Why?

tweekmonster replied:
TBH, late night brute force bug squashing.  I'll have to recheck this since I rewrote how the buffer was created after this was added.

justinmk commented:
Let the messages buffer be a normal buffer. If user deletes it, we can deal with it (recreate it, or fallback to the pager).

tweekmonster commented:
The answer to "why pane not buffer" is actually petty:  `msgbuf` is a variable that's already used in a couple of places.  TBH, I don't like "message buffer", but "history buffer" didn't sound right either.

mhinz commented:
Just "history buffer" would be confusing anyway, since there are many types of histories: message history, cmdline history, search history, ...

It just feels like "pane" would introduce a completely new word to the Vim vocabulary for no good reason. On the other hand, I can't come up with a better suggestion right now..

tweekmonster commented:
Maybe `message window` and `msgwin`?

bfredl commented:
How is the quickfix buffer implemented? is it a "normal" nofile buffer with some mappings set up?

tweekmonster commented:
> How is the quickfix buffer implemented? is it a "normal" nofile buffer with some mappings set up?

Kind of.  It's buftype is set to `quickfix` while it's filename is unset.  You can run `:file /tmp/whatever` and `:set modifiable buftype=` and it'll be a run of the mill buffer that can be edited.

tweekmonster committed:
Undo msg_silent on early return.

tweekmonster committed:
NULL check after buflist_new()

tweekmonster committed:
Check syn_name2id() return value.

tweekmonster committed:
Moved check for msgpane()'s third argument.

tweekmonster committed:
Fixed some typos.

justinmk commented:
It's not a window, it's a buffer. It could be open in multiple windows or a tab. Cannot see a better user-facing name than "message buffer".

tweekmonster commented:
But, when presented to the user, it's in a window.  It's all about using the right wording to set expectations 😉

Seriously, I agree.  My reason for the name isn't that good.  I just didn't want the other instances of `msgbuf` to be mistaken has being associated and I didn't want long names like `message_buffer_whatever()`.  Also, `msgbuf_buf`.

tweekmonster committed:
tests: Try without a window on screen.

tweekmonster committed:
Use non_zero_arg() to check msgpane()'s third argument.

tweekmonster committed:
tests: More tests.

tweekmonster committed:
Changed screen updating when not in a visible window.

Popup menu was causing the current window to update with the message
buffer's changes when it wasn't on screen.

tweekmonster committed:
Use cmdmod split when opening the message buffer.

tweekmonster commented:
Now that I've been using this for a couple days, there's a few things I think should happen:

- No lines are written to the buffer until it's in a window.
- Use a short delay timer that waits a few ms before adding lines to the buffer.
- `msgpane()` should treat each line in the input string as a line in the buffer instead of displaying it as a really long line with NUL characters.  It would also be the notable difference from `:echomsg`

justinmk commented:
Timers could add bugs. Wait on optimizations until another pr. Though collecting messages in a list would still be my recommendation... OTOH a buffer can be thought of as a list.

tweekmonster commented:
> Timers could add bugs. Wait on optimizations until another pr.

Point taken.

> Though collecting messages in a list would still be my recommendation

I do like the idea, but I still want to keep the colors and have a buffer that's updated when on screen.  The default messages only stores the highlight attr and I can't think of how to reverse it into a highlight name.

Maybe this can be exposed through `histget('messages', ...)` for scripts.

justinmk commented:
> Maybe this can be exposed through histget('messages', ...) for scripts.

Definitely seems like something we should offer, though again can wait until later.

tweekmonster committed:
Add funcs to eval.lua

tweekmonster committed:
Rename msgpane -> msgbuf, message_pane -> message_buffer

tweekmonster committed:
Force message truncation if msgbuf enabled.

tweekmonster committed:
Fixup tests

tweekmonster committed:
Don't display redirected messages

tweekmonster committed:
Lint

Shougo commented:
I think the feature should be implemented as plugin.

justinmk commented:
https://github.com/neovim/neovim/pull/27855 implements a buffer-driven pager in Lua.
